src/main/java/xonin/backhand/client/world/DummyWorld.java

```diff
@@ -70,23 +77,58 @@ public boolean updateLightByType(EnumSkyBlock p_147463_1_, int p_147463_2_, int
         return true;
     }
 
-    public Block copyAndSetBlock(World world, int x, int y, int z) {
+    @Override
+    public TileEntity getTileEntity(int x, int y, int z) {
+        if (boundBlock == null) return super.getTileEntity(x, y, z);
+        return super.getTileEntity((int) boundBlockPos.xCoord, (int) boundBlockPos.yCoord, (int) boundBlockPos.zCoord);
+    }
+
+    @Override
+    public Block getBlock(int x, int y, int z) {
+        if (boundBlock == null) return super.getBlock(x, y, z);
+        return boundBlock;
+    }
+
+    @Nullable
+    public Block copyAndSetBlock(World world, int x, int y, int z, MovingObjectPosition mop) {
+        reset();
         setBlockToAir(x, y, z);
         Block block = world.getBlock(x, y, z);
 
         if (block == null || block == Blocks.air) return null;
 
         int meta = block.getDamageValue(world, x, y, z);
-        DummyWorld.INSTANCE.setBlock(x, y, z, block, meta, 3);
+        ItemStack stack = block.getPickBlock(mop, world, x, y, z, ClientFakePlayer.INSTANCE);
+
+        if (stack == null || !stack.getItem()
+            .onItemUse(stack, ClientFakePlayer.INSTANCE, this, x, y, z, mop.sideHit, x, y, z)) {
+            setBlock(x, y, z, block, meta, 3);
+            setBoundBlock(block, x, y, z);
+        }
+
         TileEntity tile = world.getTileEntity(x, y, z);
         if (tile != null) {
             NBTTagCompound tag = new NBTTagCompound();
             tile.writeToNBT(tag);
-            TileEntity dummyTile = TileEntity.createAndLoadEntity(tag);
-            DummyWorld.INSTANCE.setTileEntity(x, y, z, dummyTile);
+            TileEntity dummyTile = getTileEntity(x, y, z);
+            if (dummyTile == null) {
+                dummyTile = TileEntity.createAndLoadEntity(tag);
+                setTileEntity(x, y, z, dummyTile);
+            }
+            dummyTile.readFromNBT(tag);
         }
 
         return getBlock(x, y, z);
     }
 
+    public void setBoundBlock(Block block, int x, int y, int z) {
+        boundBlock = block;
+        boundBlockPos = Vec3.createVectorHelper(x, y, z);
```

Comment:
nit: Any reason to make a new object instead of updating it in place?  

Not a big deal as this isn't a hot path, but excessive allocations in MC annoy me

Reply:
Fair point, I'll change it.